### DISCUSSION TITLE
Couple of questions that i am not able to find answers for

### DISCUSSION THREAD

**Discussion by ghost** [aquasecurity/trivy#7338] [Q&A]
### Question

Hello guys , 
First of all thank you very much on your efforts and on keeping this community so active and useful.

I am a new trivy user , my main usage is the MissConfig Scans ( IaC ) , trivy is so good and has a very good rules with low percentage of FP and thats why i love it.

But i am having a hard time understanding the documentations and i got couple of questions that i am not able to find answers for. 

1 -Is there a slack channel or an active chat where i can consult with people about the customized rego rules i write.

2 - Where can i find the builtin rules for K8S , Docker , cloud-formation and most importantly Terraform , this a very important request since sometimes i want to understand the rule syntax and logic, the rules of k8s and docker that were in the documeantion seems outdates since the suggestion link is (https://kubesec.io/basics/containers-securitycontext-capabilities-drop-index-all/) for example instead of aqua link https://avd.aquasec.com/misconfig/ksv003 , Is there a repo that has all of the builtin rules especially the Terraform ones ?

3 - How can i edit the policy severity / title , the only method i can think of is to exclude the builtin rule and then add a customized one by my self which is not comfortable? 

4 - is there a way not to run the built-in rule and only run the customized rules , or is there a way to download the builtin rules and run them from a config file made by me where i can choose which rules to run or where i can customize the rules how ever i want? 

5 - And again , is there any way to only run the customized rules and not the builtin rules ? ,I have thought about downloading all the config policies on a file on my system and run trivy on that config rules but it did not work , the builtin rules always runs , my goal from this question is to find a way to import the builtin rule into a file and then to customize them as much as i like 

Thank you very much 

### Target

None

### Scanner

Misconfiguration

### Output Format

None

### Mode

Standalone

### Operating System

Mac os

**Comment by nikpivkin** [comment]
Hi! Thank you for your interest.

1. https://github.com/aquasecurity/trivy?tab=readme-ov-file#community
2. All of Trivy's built-in checks are in the [trivy-checks](https://github.com/aquasecurity/trivy-checks) repository. Most of the checks are written in Go, but now we are migrating to Rego. The documentation [describes](https://aquasecurity.github.io/trivy/v0.54/docs/scanner/misconfiguration/custom/) how to write your own checks.
3. I don't think it's possible to easily modify the built-in checks
4. If you want to disable all built-in checks, the easiest way to do that is to use [Rego](https://aquasecurity.github.io/trivy/v0.54/docs/configuration/filtering/#by-rego):
```bash
$ cat filter.rego
package trivy

default ignore = false

ignore {
    startswith(input.AVDID, "AVD")
}

$ trivy conf . -d --ignore-policy filter.rego
```

If you want to enable only some of the built-in checks:
```bash
package trivy

default ignore = false

ignore {
    input.AVDID != "AVD-AWS-0090"
}
```

If you want to disable certain checks, you can use [trivyignore](https://aquasecurity.github.io/trivy/v0.54/docs/configuration/filtering/#trivyignore) or Rego.
5. [This](https://github.com/aquasecurity/trivy/discussions/2302) describes how to get all the built-in checks. Regarding filtering, see point 4.

**Comment by simar7** [comment]
hi @kariaelo - to your first question, the best place to ask would be GitHub discussions. Just like you've done now!

As for disabling builtin rules, is there a use case you are trying to achieve?

**Comment by ghost** [comment]
hi @simar7 and @nikpivkin 
First of all , thanks for your quick reply.

1- Regarding the builtin rules , are the terraform rules found here ? https://github.com/aquasecurity/trivy-checks/tree/418d48dc317381dd9ee2f84dda07db879187c20e/checks/cloud  ( Written in GO ) ?

2- " As for disabling builtin rules, is there a use case you are trying to achieve? " , I dont want to disable them and not to use them , my main goal is to run those rules from a config file on my system like a custom policies , the main goal of this is then to try adjusting them as i like , i understand from u that the only method to do this is to disable builtin rules so they won't run by deafult and then run trivy on config policy files that has all the rules that i can later modify if i want , am i correct ?

**Reply by nikpivkin** [reply]
1. List of Rego checks:
```bash
find ./checks/cloud -type f -name "*.rego" ! -name '*test.rego'
./checks/cloud/aws/codebuild/enable_encryption.rego
./checks/cloud/aws/s3/dns_compliant_name.rego
./checks/cloud/aws/s3/enable_logging.rego
./checks/cloud/aws/documentdb/encryption_customer_key.rego
./checks/cloud/aws/documentdb/enable_storage_encryption.rego
./checks/cloud/aws/documentdb/enable_log_export.rego
./checks/cloud/aws/config/aggregate_all_regions.rego
./checks/cloud/aws/accessanalyzer/enable_access_analyzer.rego
./checks/cloud/aws/dynamodb/table_customer_key.rego
./checks/cloud/aws/dynamodb/enable_recovery.rego
./checks/cloud/aws/dynamodb/enable_at_rest_encryption.rego
./checks/cloud/aws/cloudtrail/enable_log_validation.rego
./checks/cloud/aws/cloudtrail/encryption_customer_key.rego
./checks/cloud/aws/cloudtrail/no_public_log_access.rego
./checks/cloud/aws/cloudtrail/ensure_cloudwatch_integration.rego
./checks/cloud/aws/cloudtrail/enable_all_regions.rego
./checks/cloud/aws/cloudtrail/require_bucket_access_logging.rego
./checks/cloud/aws/iam/filter_iam_pass_role.rego
./checks/cloud/aws/rds/enable_iam_auth.rego
./checks/cloud/aws/rds/enable_deletion_protection.rego
./checks/cloud/aws/rds/disable_public_access.rego
./checks/cloud/aws/rds/enable_cluster_deletion_protection.rego
./checks/cloud/aws/athena/no_encryption_override.rego
./checks/cloud/aws/athena/enable_at_rest_encryption.rego
```

**Reply by ghost** [reply]
Did not understand your answer tbh , i found this folder but my question is that are the terraform rules or not ?

**Reply by nikpivkin** [reply]
We convert Terraform, CloudFormation, etc. configurations that are associated with cloud providers into a single state that is passed to the Rego input in order to use the same checks for different types of configurations. Therefore, we have no such thing as Terraform checks.

**Reply by ghost** [reply]
What about this folder https://github.com/aquasecurity/trivy-checks/tree/418d48dc317381dd9ee2f84dda07db879187c20e/checks/cloud , the policies written in GO won't work ?

**Comment by ghost** [comment]
@nikpivkin @simar7 So guys i have downloaded the builtin rules and added an ignore rego file as you stated for all of the builtin rules and i am trying to run trivy on config file with ignore policy but i get this error : 


trivy conf . --policy ~/Documents/checks --ignore-policy ~/Documents/checks/filter.rego
2024-08-14T09:47:41.653+0300	INFO	Misconfiguration scanning is enabled
2024-08-14T09:47:41.902+0300	FATAL	filesystem scan error: scan error: scan failed: failed analysis: post analysis error: post analysis error: terraform scan error: scan config error: 11 errors occurred:
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/iam/filter_iam_pass_role.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/iam/filter_iam_pass_role.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/disable_public_access.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/rds/disable_public_access.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/enable_cluster_deletion_protection.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/rds/enable_cluster_deletion_protection.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/enable_deletion_protection.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/rds/enable_deletion_protection.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/enable_iam_auth.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/rds/enable_iam_auth.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3/dns_compliant_name.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/s3/dns_compliant_name.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3/enable_logging.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/cloud/aws/s3/enable_logging.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/docker/policies/add_instead_of_copy.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/docker/add_instead_of_copy.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/docker/policies/apt_get_missing_no_install_recommends.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/docker/apt_get_missing_no_install_recommends.rego:1
Users/user/Library/Caches/trivy/policy/content/policies/docker/policies/apt_get_missing_yes_flag_to_avoid_manual_input.rego:1: rego_type_error: package annotation redeclared: Users/user/Documents/checks/docker/apt_get_missing_yes_flag_to_avoid_manual_input.rego:1
rego_compile_error: error limit reached`trivy conf . --policy ~/Documents/checks --ignore-policy ~/Documents/checks/filter.rego
2024-08-14T09:47:41.653+0300	INFO	Misconfiguration scanning is enabled
2024-08-14T09:47:41.902+0300	FATAL	filesystem scan error: scan error: scan failed: failed analysis: post analysis error: post analysis error: terraform scan error: scan config error: 11 errors occurred:
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/iam/filter_iam_pass_role.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/iam/filter_iam_pass_role.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/disable_public_access.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/rds/disable_public_access.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/enable_cluster_deletion_protection.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/rds/enable_cluster_deletion_protection.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/enable_deletion_protection.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/rds/enable_deletion_protection.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/rds/enable_iam_auth.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/rds/enable_iam_auth.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3/dns_compliant_name.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/s3/dns_compliant_name.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3/enable_logging.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/cloud/aws/s3/enable_logging.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/docker/policies/add_instead_of_copy.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/docker/add_instead_of_copy.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/docker/policies/apt_get_missing_no_install_recommends.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/docker/apt_get_missing_no_install_recommends.rego:1
Users/mostafanaa/Library/Caches/trivy/policy/content/policies/docker/policies/apt_get_missing_yes_flag_to_avoid_manual_input.rego:1: rego_type_error: package annotation redeclared: Users/mostafanaa/Documents/checks/docker/apt_get_missing_yes_flag_to_avoid_manual_input.rego:1
rego_compile_error: error limit reached`

**Reply by nikpivkin** [reply]
The package in the Rego file must be unique. Please read the documentation. https://aquasecurity.github.io/trivy/v0.54/docs/scanner/misconfiguration/custom/#package

**Reply by ghost** [reply]
I know but isnt this a builtin package ? why would it not have a unique ID ?

**Reply by ghost** [reply]
I think its conflict with the builtin rules downloaded by trivy docker

**Reply by nikpivkin** [reply]
Because Trivy still loads the built-in checks, but doesn't use them in reports.

**Reply by ghost** [reply]
is there anyway that i can overcome this ?

**Reply by ghost** [reply]
I guess the only way is this :
1- Download all of the policies in checks file here : https://github.com/aquasecurity/trivy-checks/tree/418d48dc317381dd9ee2f84dda07db879187c20e/checks ( which also contains cloudforamtion and terraform ) 
2- write a script to change the package name of all of them so i won't receive any conflicts 
3 - add the ignore rule to ignore all of the builtin rules that is downloaded automatically 

And now i have all of the builtin rules running automatically from a config file that i later can i edit.

Now my question regarding this process : 

1 - Does downloading all of the polices from the checks files guarantee me to have all of the builtin rules? 
2 - does changing the package name to a random other package name would have an affect on the scan ?

**Reply by ghost** [reply]
I have tried to edit a policy that it wont have avd at the start to see if it would only run it from config but it did not work , for example :

`package sql

import (
	"github.com/aquasecurity/trivy-checks/pkg/rules"
	"github.com/aquasecurity/trivy/pkg/iac/providers"
	"github.com/aquasecurity/trivy/pkg/iac/providers/google/sql"
	"github.com/aquasecurity/trivy/pkg/iac/scan"
	"github.com/aquasecurity/trivy/pkg/iac/severity"
	"github.com/aquasecurity/trivy/pkg/iac/state"
)

var CheckPgLogCheckpoints = rules.Register(
	scan.Rule{
		AVDID:       "GCP-0025",
		Provider:    providers.GoogleProvider,
		Service:     "sql",
		ShortCode:   "pg-log-checkpoints",
		Summary:     "Ensure that logging of checkpoints is enabled.",
		Impact:      "Insufficient diagnostic data.",
		Resolution:  "Enable checkpoints logging.",
		Explanation: `Logging checkpoints provides useful diagnostic data, which can identify performance issues in an application and potential DoS vectors.`,
		Links: []string{
			"https://www.postgresql.org/docs/13/runtime-config-logging.html#GUC-LOG-CHECKPOINTS",
		},
		Terraform: &scan.EngineMetadata{
			GoodExamples:        terraformPgLogCheckpointsGoodExamples,
			BadExamples:         terraformPgLogCheckpointsBadExamples,
			Links:               terraformPgLogCheckpointsLinks,
			RemediationMarkdown: terraformPgLogCheckpointsRemediationMarkdown,
		},
		Severity: severity.Medium,
	},
	func(s *state.State) (results scan.Results) {
		for _, instance := range s.Google.SQL.Instances {
			if instance.Metadata.IsUnmanaged() {
				continue
			}
			if instance.DatabaseFamily() != sql.DatabaseFamilyPostgres {
				continue
			}
			if instance.Settings.Flags.LogCheckpoints.IsFalse() {
				results.Add(
					"Database instance is not configured to log checkpoints.",
					instance.Settings.Flags.LogCheckpoints,
				)
			} else {
				results.AddPassed(&instance)
			}
		}
		return
	},
)
 `



and here is the filter rule : 
package trivy

default ignore = false

ignore {
    startswith(input.AVDID, "AVD")
}


and then i ran this command : trivy conf . --policy ~/Documents/checks/. --ignore-policy ~/Documents/checks/filter.go 

Any idea :O

**Reply by ghost** [reply]
any help @simar7 @nikpivkin

**Reply by ghost** [reply]
Any help ?

**Reply by nikpivkin** [reply]
Hi @kariaelo 

Custom checks can only be written in Rego. Changing the package has no effect on the result.

**Reply by ghost** [reply]
can u please answer my other questions ? @nikpivkin

**Reply by nikpivkin** [reply]
I see that you gave an example of Go checks, but as I replied, custom checks can only be written in Rego.

Can you please format the code and terminal output for readability?

**Reply by ghost** [reply]
So you say that since cloud-formation and terraform rules are in Go , we won't be able to customize them ? or are those rules also written in rego ( all of them not bunch of them ) @nikpivkin

**Reply by nikpivkin** [reply]
As I answered above: most checks are written on Go, but there's a part that's written in Rego.

**Reply by ghost** [reply]
Ok thank you , so based on your answer there is no way currently to customize the Terraform and cloudformation rules since they are written in Go and not Rego , any expecation on when they will be written in rego as the rest of the checks ?

**Reply by nikpivkin** [reply]
The process of migrating to Rego is now underway.

**Reply by ghost** [reply]
@nikpivkin Thank you very much :) thats seems great , for me to make my preparations , do you have a rough guess on when will most of the checks lets say 80% by migrated to rego? would it be ready in the next 1-2 months ? Thanks again !

**Reply by nikpivkin** [reply]
You can follow the progress [here](https://github.com/aquasecurity/trivy/issues/6930) and [here](https://github.com/aquasecurity/trivy-checks/pulls?q=is%3Apr+is%3Aopen+refactor%28checks%29)

**Reply by ghost** [reply]
one last question , why is GO not customisable , i tried to download one of the GO rules and changed its desc and ID and then ran  trivy conf . --policy ~/Documents/checks/. --ignore-policy ~/Documents/checks/filter.go 
Why won't this work ? i dont get a finding on that rule although running the builtin rule with the same logic does 
thanks thanks thanks

**Reply by simar7** [reply]
`--ignore-policy` only accepts a rego check. What you are trying to do is not possible. Can you please explain your use case to us clearly so we can better understand?
